src/server/apsystem.ts

```diff
@@ -387,6 +392,31 @@ export default class ActivityPubSystem {
     }))
   }
 
+  async performUndo (fromActor: string, activity: APActivity): Promise<void> {
+    const { actor, object } = activity
+    if (typeof object !== 'string') {
+      throw createError(400, 'Undo must point to URL of object')
+    }
+    if (typeof actor !== 'string') {
+      throw createError(400, 'Activities must contain an actor string')
+    }
+
+    const inbox = this.store.forActor(fromActor).inbox
+
+    // This throws if we haven't seen this activity before
+    const existing = await inbox.get(object)
+    if (existing.actor !== actor) {
+      throw createError(400, 'Undo can only point to activities by same author')
+    }
+    await inbox.remove(object)
+
+    // Detect if follow
+    if (existing.type === 'Follow') {
+      const followerMention = await this.actorToMention(actor, fromActor)
+      await this.removeFollower(fromActor, followerMention)
+    }
```

Comment:
sorry one last thing, could you dispatch it as an approved activity too? we're not pulling but relying on webhooks to send stuff for the moderation queue